### PR TITLE
Allow adding names with comma or hyphen #110

### LINKED ISSUE #110: Add doesn't allow adding names with comma or hyphen
The regex is too restrictive (`[\\p{Alnum} ]+`)

### PR DESCRIPTION
**Pull request by kychua**
Fixes #110 

based on #144

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@kychua submitted v1 for review.

* [v1 1/2] [Make name validation regex less restrictive](https://github.com/se-edu/addressbook-level4/pull/265/commits/edf78d3b750d58a506ba5ddb9bf6a2009fc50ad0)
* [v1 2/2] [Update test person's name to test new name regex](https://github.com/se-edu/addressbook-level4/pull/265/commits/1f9d203d8e4a7989addeb1f76271756c9fef3314)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8164678f10e07049144c98acd52fc3572ad5c9a8...1f9d203d8e4a7989addeb1f76271756c9fef3314))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/person/Name.java [hunk 1]
```diff
@@ -8,8 +8,17 @@
  */
 public class Name {
 
-    public static final String MESSAGE_NAME_CONSTRAINTS = "Person names should be spaces or alphanumeric characters";
-    public static final String NAME_VALIDATION_REGEX = "[\\p{Alnum} ]+";
+    public static final String MESSAGE_NAME_CONSTRAINTS =
+            "Person names should only contain spaces, unicode characters, numbers "
```
Comment (kychua) [inline src/main/java/seedu/address/model/person/Name.java]: "unicode characters" seems too technical for a user but I'm not sure what to replace it with

File: src/main/java/seedu/address/model/person/Name.java [hunk 2]
```diff
@@ -8,8 +8,17 @@
...
+            + "or the following punctuation marks: . , ' -";
+    public static final String NAME_VALIDATION_REGEX = "^["
```
Comment (MightyCupcakes) [inline src/main/java/seedu/address/model/person/Name.java]: Should use Java's Pattern to ensure consistency with the rest of the code base (like Parser)

Comment (MightyCupcakes) [inline src/main/java/seedu/address/model/person/Name.java]: Should use Java's Pattern to ensure consistency with the rest of the code base (like Parser)

File: src/main/java/seedu/address/model/person/Name.java (hunk 1 above)
Comment (chao1995) [inline src/main/java/seedu/address/model/person/Name.java]: @kychua The issue only requests for allowing hyphen and comma in the names, I will we only add support for these two in this PR. We can support unicode in a separate PR.

File: src/main/java/seedu/address/model/person/Name.java (hunk 2 above)
Comment (chao1995) [inline src/main/java/seedu/address/model/person/Name.java]: I agree, but we can do this in another PR.

Comment (chao1995) [inline src/main/java/seedu/address/model/person/Name.java]: I agree, but we can do this in another PR.

Comment (kychua) [inline src/main/java/seedu/address/model/person/Name.java]: Yup, since this affects other classes as well

**Comment by CanIHasReview** [comment]
# v2

@kychua submitted v2 for review.

* [v2 1/2] [Make name validation regex less restrictive](https://github.com/se-edu/addressbook-level4/pull/265/commits/c72a077758b9e4f89da6a4db990e952bb3d825b1)
* [v2 2/2] [Update test person's name to test new name regex](https://github.com/se-edu/addressbook-level4/pull/265/commits/973bb8b5f14d76cd0404f1ad9c5d4695e688946d)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/da4dc8e78218b35ccf991108acf1c56ca8492f57...973bb8b5f14d76cd0404f1ad9c5d4695e688946d)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/commit/0c857fd680e94e31714b90d9e98e2fc0fde764f9))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/person/Name.java [hunk 3]
```diff
@@ -8,9 +8,19 @@
  */
```
Comment (damithc) [inline src/main/java/seedu/address/model/person/Name.java]: Commit message:
```
The original name validation only accepts alphanumeric and whitespace
characters. This is too restrictive and rejects valid names as well,
such as non-english names, hyphenated names and names containing dots,
commas or apostrophes.
```
No need for `The original`.

Add a paragraph to explain your solution. e.g. `Update regular expression to include ...`

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 4]
```diff
@@ -397,7 +397,8 @@ public void execute_find_matchesIfAnyKeywordPresent() throws Exception {
     class TestDataHelper {
 
         Person adam() throws Exception {
-            Name name = new Name("Adam Brown");
+            // unicode format based on http://stackoverflow.com/questions/4237581/comparing-unicode-characters-in-junit
+            Name name = new Name("Adam,  o'\u00E9-\u00F6 Jr. 2");
```
Comment (damithc) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Can be squashed with other commit. No need to mention updating tests in the commit message as it is expected anyway.

**Comment by CanIHasReview** [comment]
# v3

@kychua submitted v3 for review.

* [v3 1/1] [Make name validation regex less restrictive](https://github.com/se-edu/addressbook-level4/pull/265/commits/cec8d8e53d555ab74ed768dcc5c2a3b56da1a61a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...cec8d8e53d555ab74ed768dcc5c2a3b56da1a61a)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/commit/c34622a0f28bdc85c212588b12ad3274f75ca59e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by kychua** [comment]
To fix: commit message should specify that the unicode issues (case insensitive comparison, precomposed/decomposed characters) are referring to Java's `equalsIgnoreCase` method

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 5]
```diff
@@ -399,7 +399,8 @@ public void execute_find_matchesIfAnyKeywordPresent() throws Exception {
     class TestDataHelper {
 
         Person adam() throws Exception {
-            Name name = new Name("Adam Brown");
+            // unicode format based on http://stackoverflow.com/questions/4237581/comparing-unicode-characters-in-junit
+            Name name = new Name("Adam,  o'\u00E9-\u00F6 \u0041 \u030A Jr. 2");
```
Comment (damithc) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: While you are following the selected answer from the S/O, may be using UTF-8 encoding (the other answer) is makes the test code more reader friendly? Can be done as a separate issue.

**Comment by CanIHasReview** [comment]
# v4

@kychua submitted v4 for review.

* [v4 1/1] [Make name validation regex less restrictive](https://github.com/se-edu/addressbook-level4/pull/265/commits/48ad9dc502937b86bdc430706cb4d1cfc1950246)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2221d24f2e49a9ec8a9c0d45b9dcb03694d25b0f...48ad9dc502937b86bdc430706cb4d1cfc1950246)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/35bd07ae4247e2ac7b01ff459cd6fbc5eda2d0af~1...35bd07ae4247e2ac7b01ff459cd6fbc5eda2d0af))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by kychua** [comment]
Updated commit message to specify that the unicode issues are referring to Java's `String.equalsIgnoreCase()`

**Comment by damithc** [comment]
@kychua can resolve conflicts and submit a new iteration?

**Comment by kychua** [comment]
@yamgent

Didn't the original regex already reject strings of whitespace characters?
 The name string is trimmed before validating and empty strings are rejected by the regex.

> The current regex allows space characters. However, this means that a
string of whitespace characters (e.g. "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;") are permitted.
>
> Modify the regex such that blank strings (such as " ") are no longer
allowed.
>
> An alternative would be to edit the validation function to check for
the length of the trimmed string seperately (i.e. test.trim().length()
== 0). However, that would not be elegant since we will now have two
different things doing our validation (regex and the code checking
length), which feels unnecessary compared to just using regex alone.

**Comment by yamgent** [comment]
@kychua The change in the regex was made because the unit test is testing `Name#isValidName()`, not `Name#Name()`.

The callers of the methods may be trimming the string, but `Name#isValidName()` itself does not trim the string, hence lies the problem.

**Comment by kychua** [comment]
@yamgent I see. Then why do we not trim in the validation method instead? I don't think checking length is necessary, i.e. this seems to work:
```java
System.out.println(" ".trim().matches("[\\p{Alnum} ]+"));  // prints false
```
Or is it to avoid trimming twice?

**Comment by yamgent** [comment]
@kychua That would cause another issue, which is that you can now start a name with a whitespace character. This one was not explained properly, and the regex wasn't clear enough for this requirement, so I apologies for the confusion.

```java
System.out.println(" John".matches("[\\p{Alnum} ]+"));  // prints true
```

**Comment by kychua** [comment]
@yamgent I see, thanks!

**Comment by CanIHasReview** [comment]
# v5

@kychua submitted v5 for review.

* [v5 1/1] [Make name validation regex less restrictive](https://github.com/se-edu/addressbook-level4/pull/265/commits/316d358c2da931ee20ae34324519a8a462cc6203)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f40ee42cb2e29a99c9c6b8757b64b0b0f4284d67...316d358c2da931ee20ae34324519a8a462cc6203)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/d8fabb7acdef212403c0d0e19f8195275964c562~1...d8fabb7acdef212403c0d0e19f8195275964c562))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by pyokagan** [comment]
I assume @ndt93 won't be reviewing this :-P

**Comment by pyokagan** [comment]
Stepping back a bit, why do we need to validate names? Why do we see a need to control what users can or cannot put inside the "name" field? @damithc

**Comment by damithc** [comment]
> Stepping back a bit, why do we need to validate names? Why do we see a need to control what users can or cannot put inside the "name" field of their own personal address book? @damithc

Yes, there is no real need to block certain chars from the name. A more important need is to ensure a name _can_ include any valid char (people gets upset when a software says their name is 'invalid'. :-p)  For example, the name `Lee-Roberts` should be accepted even if `-R` is a valid command switch.  So, I'm OK to remove name validation altogether and let users put anything in that field. 

On a related note, name validation is not entirely evil as it can help detect user errors  (e.g user swapping values of two fields by mistake) as long as we handle possibly invalid values in a user-friendly way (e.g. warn instead of reject)

**Comment by kychua** [comment]
>Hmm, what is the point of bringing up String#equalsIgnoreCase() here? We use
regexes for name validation, right?

Find command uses String#equalsIgnoreCase().

> On another note, this probably means that two names which look the same may not
be detected as the same by our code. (via ReadOnlyPerson#isSameStateAs()).
This means that we won't properly detect duplicate names in other languages,
although I wonder if it's worth the effort to fix that.

Just realised ReadOnlyPerson#isSameStateAs() doesn't even do caseless comparisons... but yes, same issue as FindCommand.

The precomposed/decomposed marks issue can probably be fixed by using Java's Normalizer#normalize to convert the input to standard form.

The case-insensitive comparison seems harder. The ICU library [seems to be able to handle this](http://stackoverflow.com/a/34966206) but that adds a dependency. Also, I think Pierce pointed out that in Japanese, the same word can be written in different ways so we aren't really going to be able to support other languages even if we manage to fix the issues. Is it enough if we guarantee users can find their names if they use an identical search term?

**Comment by damithc** [comment]
@kychua 
> Is it enough if we guarantee users can find their names if they use an identical search term?

What's the current situation for English names and non-English names?

**Comment by kychua** [comment]
@damithc 
> What's the current situation (as per current state of this PR) for English names and non-English names?

English names work for case insensitive comparisons. Non-English names might or might not work for case-insensitive searches. Case sensitive searches should work fine if the exact same input method is used, but not if one uses precomposed characters and the other uses decomposed characters.

**Comment by damithc** [comment]
> English names work for case insensitive comparisons. Non-English names might or might not work for case-insensitive searches. Case sensitive searches should work fine if the exact same input method is used, but not if one uses precomposed characters and the other uses decomposed characters.

OK. Probably good enough for this stage.

**Comment by CanIHasReview** [comment]
# v6

@kychua submitted v6 for review.

* [v6 1/1] [Make name validation regex less restrictive](https://github.com/se-edu/addressbook-level4/pull/265/commits/db116b1784c614dafd114c5e95da84bd5802b1c6)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/65fdb4818f191c27f46846588172f644483718af...db116b1784c614dafd114c5e95da84bd5802b1c6)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/2ffa5e19fbfa97074dc6bd7a835c46de05b7e87a~1...2ffa5e19fbfa97074dc6bd7a835c46de05b7e87a))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
@kychua is the long commit message still relevant?

**Comment by kychua** [comment]
@damithc 

> is the long commit message still relevant?

I can make it more concise but the problems with unicode still exist because we are still allowing unicode characters...

File: src/main/java/seedu/address/model/person/Name.java [hunk 6]
```diff
@@ -8,15 +8,11 @@
...
-    /*
-     * The first character of the address must not be a whitespace,
-     * otherwise " " (a blank string) becomes a valid input.
+    public static final String MESSAGE_NAME_CONSTRAINTS = "Person names can take any value and should not be blank";
+    /**
+     * Names must start with a non-whitespace character. This also implies that names must not be blank e.g. "  ".
      */
-    public static final String NAME_VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
-
+    public static final String NAME_VALIDATION_REGEX = "[^\\s].*";
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Name.java]: At this point, we could just use simple string operations instead of regex?

**Comment by pyokagan** [comment]
> @damithc How do you want to proceed with this?

Forgot to put my personal opinion, I think `TODO` comments are fine, given that we aren't particularly mission critical :-P

**Comment by damithc** [comment]
> I think TODO comments are fine, given that we aren't particularly mission critical :-P

Agreed, a `TODO` is enough for now. Also agree that it's good to document 'known issues' too. May be as an appendix to the User Guide?

**Comment by pyokagan** [comment]
@kychua I just realized: while it is a good idea to be as permissive as possible wrt to characters in names, there are still some characters that we should not allow. For example, newlines, since names with newlines probably won't render correctly in our UI. So maybe keeping a list of illegal characters (maybe via a regex) would be a good idea?

**Comment by pyokagan** [comment]
@kychua Oh, to add on, I think something better than just writing `TODO` is to have failing test cases that prove that the bugs hypothesized in the commit message do actually exist.

**Comment by CanIHasReview** [comment]
# v7

@kychua submitted v7 for review.

* [v7 1/3] [Name: change validation to reject only blank names](https://github.com/se-edu/addressbook-level4/pull/265/commits/2af13c05f485c3801c54f118608d527d6aca5b3a)
* [v7 2/3] [Name: remove line terminators in name before validating](https://github.com/se-edu/addressbook-level4/pull/265/commits/32f962372633f9005b3df06cfa0014b3ca9f913b)
* [v7 3/3] [Address: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/265/commits/3b9aaac273f0b3c3ac9f005c0e13dddf5febb595)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/297506b87fabe6de6baa7901388039ec50199802...3b9aaac273f0b3c3ac9f005c0e13dddf5febb595)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/compare/d16bfe4d89f0c84d5d18ecbf571d5131d712bdd2~1...d16bfe4d89f0c84d5d18ecbf571d5131d712bdd2))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v8

@kychua submitted v8 for review.

* [v8 1/4] [testutil: Add ExpectedFailure annotation for tests](https://github.com/se-edu/addressbook-level4/pull/265/commits/5a4f4d396c909c94465aa0a9edb2db23bea24895)
* [v8 2/4] [Name: change validation to reject only blank names](https://github.com/se-edu/addressbook-level4/pull/265/commits/9af2a1637874f785b2ce27b749f5da8f30e0c418)
* [v8 3/4] [Name: remove line terminators in name before validating](https://github.com/se-edu/addressbook-level4/pull/265/commits/5201cf3714a61d9e42a5f426dcd2a65d93d15fe7)
* [v8 4/4] [Address: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/265/commits/9bcd3869de1af97cf67c7c58fd8199365421ff54)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/e1eafd1145cb5fdcc81bbacecf9e5145903b3fe2...9bcd3869de1af97cf67c7c58fd8199365421ff54)) ([:chart_with_upwards_trend: Interdiff between v7 and v8](https://github.com/se-edu/addressbook-level4/compare/e8918b5af79b1a5030720bed67bdbca1485f6b2b~1...e8918b5af79b1a5030720bed67bdbca1485f6b2b))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/265/8/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by kychua** [comment]
@pyokagan 
> while it is a good idea to be as permissive as possible wrt to characters in names, there are still some characters that we should not allow. For example, newlines, since names with newlines probably won't render correctly in our UI. So maybe keeping a list of illegal characters (maybe via a regex) would be a good idea?

I stripped removed line breaks instead of marking multi-line names invalid since JavaFx strips the newline character. Can't think of what other characters to ban...

**Comment by kychua** [comment]
# v8

* Added ExpectedFailure annotation to mark tests -> by right it should be shown somewhere that these tests failed but that it's expected, but I don't know how/where to show it
* Added tests to log bugs
* Added `StringUtil#isBlank(String)` to support unicode whitespace instead of using `str.trim().isEmpty()`
* Handle multi-line names by removing line breaks
* Update `Address` validation to use `StringUtil#isBlank(String)` instead of regex

**Comment by Zhiyuan-Amos** [comment]
Should we close this?

**Comment by eugenepeh** [comment]
Think we can put it on hold till @kychua decides to resume it

**Comment by Zhiyuan-Amos** [comment]
@kychua do you plan to resume on it? :P

**Comment by kychua** [comment]
@Zhiyuan-Amos 
> do you plan to resume on it?

nope, I won't be working on this anymore.
